photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.function
+
+import breeze.linalg.Vector
+import org.apache.spark.broadcast.Broadcast
+
+import com.linkedin.photon.ml.normalization.NormalizationContext
+
+/**
+ * The base objective function class for an optimization problem.
+ */
+abstract class ObjectiveFunction {
+  type Data
+  type Coefficients
+
```

Comment:
@XianXing I have not forgotten this question! As part of InDay today, I went through and tried to replace abstract types with generics. I was able to do so - it looks like my previous errors were due to the half-done state of the code in between two simultaneous refactors. However, I still think that abstract types is the way to go. Please take a look at my personal fork - you should be able to see a branch called `xian` which has generics instead.

I feel that the abstract types make the code less cluttered. Yes, in some places the types are implicit instead of explicit, but this is only in locations where the exact types are not important: the relationships between the classes are. The types are kept explicit in the locations where they are bound.

Please let me know what you think, as always I value you input.